### PR TITLE
Implement summarize selected text in context menu

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/35640
Resolves https://github.com/brave/brave-browser/issues/35543
Resolves https://github.com/brave/brave-browser/issues/35355

A new setting which is default on:
Screenshot 2024-01-30 at 5 24 55 PM

user prompt - light
Screenshot 2024-01-30 at 2 03 55 PM

user prompt - dark
Screenshot 2024-01-30 at 2 03 38 PM

Updated opt-in message:
Screenshot 2024-02-06 at 9 33 42 AM

https://github.com/brave/brave-core/assets/4730197/26d41ddf-1f2d-4b90-a5b2-afa208a6c425

Page content will not be sent if Leo side panel is not open and there is no existing conversations.
Otherwise it will follow the current state.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1. Make sure all scenarios listed in https://github.com/brave/brave-browser/issues/35640#issue-2104020314 is working as expected.
2. When the new setting is enabled, make sure Leo tools shows up when selected text in a page.
3. When the new setting is disabled, make sure Leo tools does not show up when selected text in a page.
4. Clear Leo data and try to summarize selected text via context menu should show the opt-in message.

### CODE REVIEW

File: components/resources/ai_chat_ui_strings.grdp [hunk 1]
```diff
@@ -36,6 +36,9 @@
   <message name="IDS_CHAT_UI_SUMMARIZE_PAGE" desc="Button to summarize the article on the current page">
     Summarize this page
   </message>
+  <message name="IDS_CHAT_UI_SUMMARIZE_SELECTED_TEXT" desc="Text for summarizing selected text">
+    Summarize this excerpt
```
Comment (mkarolin) [inline components/resources/ai_chat_ui_strings.grdp:40]: Nit: "Summarize selection" might be simpler, but no strong preference.

**Review by mkarolin — APPROVED** [review]
`chromium_src`++
`strings`++

File: app/brave_settings_strings.grdp [hunk 2]
```diff
@@ -1439,6 +1439,9 @@
   <message name="IDS_SETTINGS_LEO_ASSISTANT_SHOW_ICON_ON_TOOLBAR_LABEL" desc="The text for settings option">
     Show Leo icon in the sidebar
   </message>
+  <message name="IDS_SETTINGS_LEO_ASSISTANT_SHOW_IN_CONTEXT_MENU_LABEL" desc="The text for settings option">
+    Show Leo in the context menu
```
Comment (petemill) [inline app/brave_settings_strings.grdp:1443]: I can't find this in the design, but to me it seems like it should be something more explanative, maybe even just a small change like "Show Leo tools in the context menu [on websites]"?

File: components/ai_chat/core/browser/conversation_driver.cc [hunk 3]
```diff
@@ -598,6 +599,25 @@ void ConversationDriver::OnSuggestedQuestionsResponse(
...
+  // When panel is open or has existing conversation, do not change the state.
+  return !is_conversation_active_ && chat_history_.empty();
+}
+
+void ConversationDriver::SummarizeSelectedText(
+    const std::string& selected_text,
+    bool should_unlink_page_content) {
+  if (should_unlink_page_content) {
+    SetShouldSendPageContents(false);
+  }
```
Comment (petemill) [inline components/ai_chat/core/browser/conversation_driver.cc:613]: Having this as a param seems inconsistent when the other similar public functions don't. I understand that you've added `ShouldUnlinkPageContent` to make sure that page content isn't linked mysteriously from "external" requests. At the moment these requests are from 1) context menu and 2) omnibox.

Omnibox is calling `SubmitHumanConversationEntry` which doesn't have the `should_unlink_page_content` param, and context menu is calling `SummarizeSelectedText` which does.

Perhaps we can make this a little more readable and consistent. How about we contain this logic in the same place? Perhaps ConversationDriver can simply unlink the page content if the conditions in `ShouldUnlinkPageContent` are `true` and we can make `ShouldUnlinkPageContent` a private member?

In other words, `SubmitHumanConversationEntry` could have the following inserted at the beginning:

```cpp
if (ShouldUnlinkPageContent()) {
  SetShouldSendPageContents(false);
}
```
or do this logic inline, or if it has to happen in multiple places, change the name of `ShouldUnlinkPageContents` to `MaybeUnlinkPageContents` and it can perform the decision.

File: components/ai_chat/resources/page/components/quote/style.module.scss [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+.quote .line {
+  flex: none;
+  order: 0;
+  align-self: stretch;
+  flex-grow: 0;
+}
+
+.quote .text {
```
Comment (petemill) [inline components/ai_chat/resources/page/components/quote/style.module.scss:36]: `.text` (and `.line`) are always inside `.quote`, so there's no need to split the styles up for them by defining them in two places. They can all be simply under `.text` (or `.line`)

File: components/ai_chat/resources/page/components/quote/index.tsx [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import styles from './style.module.scss'
+
+interface QuoteProps {
+  text: string
+}
+
+function Quote (props: QuoteProps) {
+  return (
+    <div className={styles.quote}>
+      <div className={styles.line}></div>
```
Comment (petemill) [inline components/ai_chat/resources/page/components/quote/index.tsx:16]: Looks like this line element exists because the design for this line has a fully rounded "nose". Since it's a non-content element, it would be great to have it be a pseudo element defined only in css. It can be `.text::before { ...`

**Review by petemill — CHANGES_REQUESTED** [review]
Just a couple of very minor code suggestions. I reviewed most of the code but not the tests deeply yet, I'll do that next.

In the meantime, there's a couple of differences from the intended UI:

1. In this PR, we're showing the prompt we give to the engine as if the user has typed it out. In other words, we're putting words in the user's mouth 😀. The summarize action is meant to have a kind of label. To achieve this we could have `ConversationTurn` have a field (e.g. `ActionType`) with values `Text | SummarizeSelection | ...`. This could help with both labelling and any custom display or follow-up action we want to be available in the UI.

    If you want, we can do this as a follow-up and include the regular "summarize this page" button for that custom labelling. (IMO that button should be special compared to the regular suggested questions and call the `SubmitSummarizationRequest` function).

    image


3. This PR adds a simple right-click option with the string "Summarize". It doesn't mention Leo. Just checking whether you've discussed that change with anyone? In the design it's in a "Leo tools" submenu:

image

File: components/ai_chat/resources/page/components/quote/style.module.scss (hunk 4 above)
Comment (yrliou) [reply components/ai_chat/resources/page/components/quote/style.module.scss]: fixed in https://github.com/brave/brave-core/pull/21670/commits/d553512aa99fa9b340158c91058f506414c2816d

File: components/ai_chat/core/browser/conversation_driver.cc (hunk 3 above)
Comment (yrliou) [reply components/ai_chat/core/browser/conversation_driver.cc]: fixed by https://github.com/brave/brave-core/pull/21670/commits/fc4a6dd8f6dd5a95d867e959c1a17d46a19b8712
Can't put inside `SubmitHumanConversationEntry` as the panel activation status already changed when it's called.

File: app/brave_settings_strings.grdp (hunk 2 above)
Comment (yrliou) [reply app/brave_settings_strings.grdp]: fixed in https://github.com/brave/brave-core/pull/21670/commits/d553512aa99fa9b340158c91058f506414c2816d

**Comment by yrliou** [comment]
Updated context menu:
Screenshot 2024-01-29 at 9 12 26 PM

I tried to add Leo logo too in this PR but it doesn't show up when I try to do
`menu_model_.SetIcon(index, ui::ImageModel::FromVectorIcon(kLeoProductBraveLeoIcon, kColorIcon, 16);`
I'm not sure why but I would like to figure this out later in a follow-up.

I'll also create a follow-up for below, I wasn't sure how to do this new label stuff in UI so I just did the question style in the first version.
Screenshot 2024-01-29 at 9 19 26 PM

File: components/ai_chat/resources/page/components/quote/index.tsx (hunk 5 above)
Comment (yrliou) [reply components/ai_chat/resources/page/components/quote/index.tsx]: fixed in https://github.com/brave/brave-core/pull/21670/commits/8d21c79a4791598fd535ca7a855718e035255cd5, let me know if other CSS changes are needed still.

**Comment by yrliou** [comment]
I do have `ActionType` when I was developing but decided to leave it out and make it consistent with existing code for now, I could add it back.

File: components/ai_chat/resources/page/components/privacy_message/index.tsx [hunk 6]
```diff
@@ -16,24 +16,33 @@
...
+  const handleLinkClick = (url: string) => {
     const mojomUrl = new Url()
-    mojomUrl.url = WIKI_URL
+    mojomUrl.url = url
 
     getPageHandlerInstance().pageHandler.openURL(mojomUrl)
   }
 
+  const createLinkWithClickHandler = (content: string, url: string) => (
+      <a onClick={() => handleLinkClick(url)} href={url} target='_blank'>
```
[the remark this answers is not in the document]
Comment (petemill) [reply components/ai_chat/resources/page/components/privacy_message/index.tsx:32]: It's safe from WebUI

File: components/ai_chat/core/browser/engine/engine_consumer_claude.cc [hunk 7]
```diff
@@ -98,7 +116,8 @@ std::string BuildClaudePrompt(
   std::string date_and_time_string =
       base::UTF16ToUTF8(TimeFormatFriendlyDateAndTime(base::Time::Now()));
   std::string prompt = base::StrCat(
-      {kHumanPromptSequence, prompt_segment_article,
+      {kHumanPromptSequence, prompt_segment_selected_text,
+       prompt_segment_article,
```
Comment (petemill) [inline components/ai_chat/core/browser/engine/engine_consumer_claude.cc:120]: It seems like we're connecting the latest question to the selected text in the wrong part of the prompt - this is the system part - the introduction before the conversation history and well before the latest/current human question. The way we're doing it in `GetConversationHistoryString` seems most appropriate - why not do it like that also for the latest question?
Otherwise it seems like the whole conversation is being tied to that excerpt and that might not lead to the AI restricting the answer to that excerpt especially when there is already a conversation. Or am I missing something?

File: components/ai_chat/core/browser/engine/engine_consumer_claude_unittest.cc [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+       "Which show is this catchphrase from?", "I have spoken."},
+      {mojom::CharacterType::ASSISTANT, mojom::ActionType::RESPONSE,
+       mojom::ConversationTurnVisibility::VISIBLE, "The Mandalorian.",
+       std::nullopt}};
+  auto* mock_remote_completion_client =
+      static_cast<MockRemoteCompletionClient*>(engine_->GetAPIForTesting());
+  std::string prompt_before_time_and_date =
+      "\n\nHuman: Here is the excerpt of the content in <excerpt> "
+      "tags:\n<excerpt>\nI'm groot.\n</excerpt>\n\nA user selects this "
+      "excerpt from the page content.\n\nHere is the text of a web page in "
```
Comment (petemill) [inline components/ai_chat/core/browser/engine/engine_consumer_claude_unittest.cc:66]: This is the part that I think should be located to just before "Here is the user's question" part, and possibly change it to "Here is the user's question about that excerpt"

**Review by petemill — COMMENTED** [review]
Frontend and UI looks good now. 
GREAT work on the testing and mock API implementation.
Just a question on the prompts - seems backwards to me, but maybe you discussed with @lorenzominto already? (same applies to llama/mixtral prompts)

File: components/ai_chat/core/browser/engine/engine_consumer_claude_unittest.cc (hunk 8 above)
Comment (yrliou) [reply components/ai_chat/core/browser/engine/engine_consumer_claude_unittest.cc:66]: Thank you for pointing out, was trying to get it adjusted during the review, I'll point it out directly or ask @LorenzoMinto  first next time. Asked him in the DM now, will adjust.

**Review by Brandon-T — APPROVED** [review]

File: components/ai_chat/core/browser/engine/engine_consumer_claude_unittest.cc (hunk 8 above)
Comment (LorenzoMinto) [reply components/ai_chat/core/browser/engine/engine_consumer_claude_unittest.cc:66]: Agree with @petemill. We should move the EXCERPT bit below the history and above the user question, and use an excerpt-specific question 👍

File: components/resources/ai_chat_prompts.grdp [hunk 9]
```diff
@@ -62,6 +71,12 @@ Use unicode symbols for formatting where appropriate. Use backticks (`) to wrap
 <message name="IDS_AI_CHAT_LLAMA2_GENERAL_SEED" translateable="false">
 Here is your response:</message>
 
+<message name="IDS_AI_CHAT_LLAMA2_SELECTED_TEXT_PROMPT_SEGMENT" translateable="false">
+This is the excerpt of the content in &lt;excerpt&gt; tags:
```
Comment (LorenzoMinto) [inline components/resources/ai_chat_prompts.grdp:75]: nit: `page content` since we are using `<page>` tags

File: components/ai_chat/core/browser/engine/engine_consumer_claude_unittest.cc (hunk 8 above)
Comment (LorenzoMinto) [reply components/ai_chat/core/browser/engine/engine_consumer_claude_unittest.cc:66]: On a second thought, since excerpts might serve as context for subsequent requests and because it might be the only context available (where no page content is added), I think it might be better to have it as an addendum to the system prompt, but after the page content segment. It might still be worth changing the question formulation to reflect whether <excerpt>/<page> are available. Wdyt @petemill?

Another question, could there be more than one excerpt per conversation?

Comment (yrliou) [reply components/ai_chat/core/browser/engine/engine_consumer_claude_unittest.cc:66]: It could be possible in a conversation, there’re other previous selected text in the history, but they're not in excerpt tag anymore, only the latest one will be inside tag.

**Comment by yrliou** [comment]
Prompt is updated based on @LorenzoMinto's current suggestion.
Claude:
Screenshot 2024-02-01 at 1 36 19 PM

For llama/mixtral, only update the `IDS_AI_CHAT_LLAMA2_SELECTED_TEXT_PROMPT_SEGMENT` to start with
`This is the excerpt of the content in &lt;excerpt&gt; tags:`.

**Comment by petemill** [comment]
> Prompt is updated based on @LorenzoMinto's current suggestion. Claude:
> 
> * Move excerpt after page content but still before history.
> * Use `Here is the user's question about the excerpt` for question part.
> * Minor text updates like "Here is an excerpt of the page content".
> 
> Screenshot 2024-02-01 at 1 36 19 PM
> For llama/mixtral, only update the `IDS_AI_CHAT_LLAMA2_SELECTED_TEXT_PROMPT_SEGMENT` to start with `This is an excerpt of the page content in &lt;excerpt&gt; tags:`.

@LorenzoMinto what do you think about this? I think in this example that the selected text is so watered down by the time you get to the question that I don't think it will restrict itself. What's the advantage of separating the question from the question's context? (This is part of the reason I suggested ignoring history for these "commands", but with history it seems there's more reason to co-locate the question with the selected text).

Why not simply:
1. (optional, if connected) text of page
2. system prompt
3. conversation history, including selected text for each entry if applicable, wrapped in `<history>`
    e.g. `Summarize the following excerpt: <excerpt>some text</excerpt>`
5. current question, including selected text if applicable
    e.g. `Summarize the following excerpt: <excerpt>some text</excerpt>`

I don't think follow-up considerations are a concern because a follow-up question itself would not have selected text, so it won't appear at the beginning of the prompt anyway in the current state of the PR.

**Comment by LorenzoMinto** [comment]
I agree with having the excerpt near the question for task like Summarisation and One-off questions (the prompt structure you suggested). That will keep the context focused on the excerpt. But from my understanding of the UI/UX, also asking @yrliou, One-off questions are already follow-ups on the summarization. Is there a way to ask the question directly without summarizing it first?

The goal of having a space for excerpts at the top of the prompt was to keep the excerpt context clear for follow-ups, then we target the question to the excerpt by specifying that it's about it. That said, I agree it will wash down the focus a bit, but Claude should handle it well. Haven't tested the functionality myself yet.

There is also the case where there is no page content but only the excerpt, so we might want to keep that in its place for longer.

**Comment by LorenzoMinto** [comment]
TLDR: If the scope of this PR is one-off usage: summarize(/check/rephrase) an excerpt, then we should go @petemill's suggestion. We can make the prompt structure more complex if we need more follow-up functionality.

**Comment by yrliou** [comment]
Prompt update after https://github.com/brave/brave-core/pull/21670/commits/735e84e821ce6907c67c319fbcd023bb03a22ed7
```
Human: Here is the text of a web page in <page> tags:
<page>...</page>

A user is reading this web page.

[ system prompt ]

Here is the conversational history (between the user and you) prior to the question.
<history>...</history>

Here is an excerpt of the page content in <excerpt> tags:
<excerpt>...</excerpt>

A user selects this excerpt from the page content.

Here is the user's question about the excerpt:
<question>
Summarize the excerpt
</question>

How do you respond to the user's question? Put your response in <response></response> tags.

Assistant:  <response>
```

**Comment by yrliou** [comment]
@LorenzoMinto @petemill Please take a look at the latest, thanks.

File: components/resources/ai_chat_prompts.grdp [hunk 10]
```diff
@@ -30,6 +41,14 @@ How do you respond to the user's question? Put your response in &lt;response&gt;
 
 A user is reading this web page.</message>
 
+<message name="IDS_AI_CHAT_CLAUDE_SELECTED_TEXT_PROMPT_SEGMENT" translateable="false">Here is an excerpt of the page content in &lt;excerpt&gt; tags:
+&lt;excerpt&gt;
+<ph name="EXCERPT">$1</ph>
+&lt;/excerpt&gt;
+
+A user selects this excerpt from the page content.
```
Comment (petemill) [inline components/resources/ai_chat_prompts.grdp:49]: "A user" -> "The user"

**Review by petemill — APPROVED** [review]

File: components/ai_chat/core/browser/conversation_driver_unittest.cc [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+        EXPECT_NE(body.find("I have spoken."), std::string::npos);
+        EXPECT_NE(body.find(l10n_util::GetStringUTF8(
+                      IDS_AI_CHAT_QUESTION_SUMMARIZE_SELECTED_TEXT)),
+                  std::string::npos);
+        if (is_page_content_linked) {
+          EXPECT_NE(body.find("The child's name is Grogu."), std::string::npos);
+        }
+        // Set header for enabling SSE.
+        auto head = network::mojom::URLResponseHead::New();
+        head->mime_type = "text/event-stream";
```
Comment (petemill) [inline components/ai_chat/core/browser/conversation_driver_unittest.cc:136]: by the way, we can set a featureparam to force SSE disabled, but this seems simple enough for now

Comment (yrliou) [reply components/ai_chat/core/browser/conversation_driver_unittest.cc:136]: Ideally we should test both cases in the end, I choose this as a starter as that's what my env was using.
Comprehensive unit test coverage of ConversationDriver is out of the PR scope, so here is just a starter for adding tests.

File: components/resources/ai_chat_prompts.grdp (hunk 10 above)
Comment (yrliou) [reply components/resources/ai_chat_prompts.grdp:49]: @petemill quick question, in article prompt, it is `A user is reading this web page.`, why the difference?

Comment (yrliou) [reply components/resources/ai_chat_prompts.grdp:49]: same goes for video prompt segment, it is currently `A user is watching this video.`

Comment (LorenzoMinto) [reply components/resources/ai_chat_prompts.grdp:49]: using `The` might avoid confusion for the LLM that it's the same user doing the selection, the first `A user` kind of sets the stage.

**Review by LorenzoMinto — APPROVED** [review]
lgtm++

File: components/resources/ai_chat_prompts.grdp (hunk 10 above)
Comment (yrliou) [reply components/resources/ai_chat_prompts.grdp]: Updated, thanks!

**Comment by yrliou** [comment]
Follow up for custom labelling created:  https://github.com/brave/brave-browser/issues/35842

**Comment by kjozwiak** [comment]
**`Quick Note:`** The following basically checks the `opt-in` copy changes that are being uplifted into `1.63.x` via https://github.com/brave/brave-core/pull/21943 and **doesn't** verify the various issues being fixed via https://github.com/brave/brave-core/pull/21670#issue-2093097604.

Verification **PASSED** on `Win 11 x64` using the following build(s):

```
Brave | 1.64.55 Chromium: 121.0.6167.164 (Official Build) nightly (64-bit)
-- | --
Revision | b6025f4892ad4674e751eccbb6e61c9aa7bbd1c4
OS | Windows 11 Version 23H2 (Build 22631.3085)
```

As per https://github.com/brave/brave-browser/issues/35355#issue-2084905331, ensured that the title is changed to `Chat Privately with Brave Leo` and the CTA button is changed to `I understand` as per the following:

leoStringChange

Verification **PASSED** on `Pixel 6` running `Android 14` using the following build(s):

```
Brave | 1.64.55 Chromium: 121.0.6167.164 (Official Build) canary (64-bit)
--- | ---
Revision | b6025f4892ad4674e751eccbb6e61c9aa7bbd1c4
OS | Android 14; Build/AP11.231215.009; 34; REL
```

As per https://github.com/brave/brave-browser/issues/35355#issue-2084905331, ensured that the title is changed to `Chat Privately with Brave Leo` and the CTA button is changed to `I understand` as per the following:

Screenshot_20240208-142608
